### PR TITLE
Make the two Equals implementations consistent

### PR DESCRIPTION
**Pull request by Therzok**
Simplify the object variant and make them use ~same code.

The object variant was querying IsDirectBinding on `this` twice and null checking obj twice.

### CODE REVIEW

File: src/Foundation/NSObject2.cs [hunk 1]
```diff
@@ -734,25 +734,24 @@ public override int GetHashCode ()
...
 		public override bool Equals (object obj)
 		{
-			if (obj == null)
-				return false;
 			var o = obj as NSObject;
 			if (o == null)
 				return false;
-			// is only one is a direct binding then both cannot be equals
-			if (IsDirectBinding != o.IsDirectBinding)
-				return false;
```
[the remark this answers is not in the document]
Comment (spouliot) [reply src/Foundation/NSObject2.cs]: I wonder if a simple `return (obj as NSObject);` is not a better.
There's almost no chance for the linker to eliminate the `IEquatable<T>` implementation anyway.

[the remark this answers is not in the document]
Comment (spouliot) [reply src/Foundation/NSObject2.cs]: It's a bit more complex (test failures)
`IEquatable<T>` support was added (well after the original) `NSObject`
so we wanted to be sure any existing code that override `Equals(object)` would continue to behave the same as before (that's what the failing test is checking). That's why we end up calling the existing `Object` virtual.

[the remark this answers is not in the document]
Comment (Therzok) [reply src/Foundation/NSObject2.cs]: The check is still there though!

File: src/Foundation/NSObject2.cs [hunk 2]
```diff
@@ -734,26 +734,17 @@ public override int GetHashCode ()
...
+				
 			// we can only ask `isEqual:` to test equality if both objects are direct bindings
-			return IsDirectBinding ? IsEqual (o) : Object.ReferenceEquals (this, obj);
+			bool isDirectBinding = IsDirectBinding;
+			return (isDirectBinding == o.IsDirectBinding) && (isDirectBinding ? IsEqual (o) : ReferenceEquals (this, o));
```
Comment (Therzok) [inline src/Foundation/NSObject2.cs]: The IsDirectBinding equality check is here.

Comment (spouliot) [reply src/Foundation/NSObject2.cs]: yes, it just a readability comment :)

Comment (Therzok) [reply src/Foundation/NSObject2.cs]: Will fix!

**Comment by spouliot** [comment]
> Failed to compare API and create generator diff
is unrelated and fixed in another PR
Everything else is green

**Pull request merged**